extensions/notebook/src/book/bookTreeView.ts

```diff
@@ -6,101 +6,86 @@
+import CodeAdapter from '../prompts/adapter';
+import { BookTreeItem } from './bookTreeItem';
 import * as nls from 'vscode-nls';
+import { isEditorTitleFree } from '../common/utils';
 import { promisify } from 'util';
-import { IJupyterBookToc, IJupyterBookSection } from '../contracts/content';
+import { BookModel } from './bookModel';
 
 const localize = nls.loadMessageBundle();
-const existsAsync = promisify(fs.exists);
+const exists = promisify(fs.exists);
 
-export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeItem>, azdata.nb.NavigationProvider {
-	readonly providerId: string = 'BookNavigator';
+export class BookTreeViewProvider implements vscode.TreeDataProvider<BookTreeItem> {
 
 	private _onDidChangeTreeData: vscode.EventEmitter<BookTreeItem | undefined> = new vscode.EventEmitter<BookTreeItem | undefined>();
 	readonly onDidChangeTreeData: vscode.Event<BookTreeItem | undefined> = this._onDidChangeTreeData.event;
-	private _tableOfContentPaths: string[] = [];
-	private _allNotebooks = new Map<string, BookTreeItem>();
-	private _extensionContext: vscode.ExtensionContext;
 	private _throttleTimer: any;
 	private _resource: string;
+	private _extensionContext: vscode.ExtensionContext;
+	private prompter: IPrompter;
 	// For testing
 	private _errorMessage: string;
 	private _onReadAllTOCFiles: vscode.EventEmitter<void> = new vscode.EventEmitter<void>();
 	private _openAsUntitled: boolean;
+	public viewId: string;
+	public books: BookModel[];
+	public currentBook: BookModel;
+
+	constructor(workspaceFolders: vscode.WorkspaceFolder[], extensionContext: vscode.ExtensionContext, openAsUntitled: boolean, view: string) {
+		this._openAsUntitled = openAsUntitled;
+		this._extensionContext = extensionContext;
+		this.books = [];
+		this.initialize(workspaceFolders, null);
+		vscode.commands.executeCommand('setContext', 'untitledBooks', openAsUntitled);
+		this.viewId = view;
+		this.prompter = new CodeAdapter();
 
-	constructor(workspaceFolders: vscode.WorkspaceFolder[], extensionContext: vscode.ExtensionContext) {
-		this.initialize(workspaceFolders, null, extensionContext);
 	}
 
-	private initialize(workspaceFolders: vscode.WorkspaceFolder[], bookPath: string, context: vscode.ExtensionContext): void {
-		let workspacePaths: string[] = [];
+	private async initialize(workspaceFolders: vscode.WorkspaceFolder[], bookPath: string): Promise<void> {
 		if (bookPath) {
-			workspacePaths.push(bookPath);
+			let book: BookModel = new BookModel(bookPath, this._openAsUntitled, this._extensionContext);
+			await book.initializeContents();
+			this.books.push(book);
+			if (!this.currentBook) {
+				this.currentBook = book;
+			}
 		}
 		else if (workspaceFolders) {
-			workspacePaths = workspaceFolders.map(a => a.uri.fsPath);
+			workspaceFolders.map(async (a) => {
+				let book: BookModel = new BookModel(a.uri.fsPath, this._openAsUntitled, this._extensionContext);
```

Comment:
You should wrap this in a Promise.all - currently you're just returning immediately while the async methods run separately